ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -0,0 +1,229 @@
+{-# LANGUAGE RankNTypes #-}
+module Development.IDE.Core.ProgressReporting
+  ( ProgressEvent(..)
+  , ProgressReporting(..)
+  , noProgressReporting
+  , delayedProgressReporting
+  , directProgressReporting
+  -- utilities, reexported for use in Core.Shake
+  , mRunLspT
+  , mRunLspTCallback
+  ) where
+
+import           Control.Concurrent.Async
+import           Control.Concurrent.STM
+import           Control.Concurrent.Strict
+import           Control.Monad.Extra
+import           Control.Monad.IO.Class
+import qualified Control.Monad.STM              as STM
+import           Control.Monad.Trans.Class      (lift)
+import           Data.Foldable                  (for_)
+import           Data.HashMap.Strict            (HashMap)
+import qualified Data.HashMap.Strict            as HMap
+import           Data.IORef
+import qualified Data.Text                      as T
+import           Data.Unique
+import           Development.IDE.GHC.Orphans    ()
+import           Development.IDE.Graph          hiding (ShakeValue)
+import           Development.IDE.Types.Location
+import           Development.IDE.Types.Options
+import           GHC.IORef                      (atomicModifyIORef'_,
+                                                 atomicSwapIORef)
+import qualified Language.LSP.Server            as LSP
+import           Language.LSP.Types
+import qualified Language.LSP.Types             as LSP
+import           System.Time.Extra
+import           UnliftIO.Exception             (bracket_)
+
+data ProgressEvent
+    = KickStarted
+    | KickCompleted
+
+data ProgressReporting  = ProgressReporting
+  { progressUpdate :: ProgressEvent -> IO ()
+  , inProgress     :: forall a. NormalizedFilePath -> Action a -> Action a
+  , progressStop   :: IO ()
+  }
+
+noProgressReporting :: IO ProgressReporting
+noProgressReporting = return $ ProgressReporting
+  { progressUpdate = const $ pure ()
+  , inProgress = const id
+  , progressStop   = pure ()
+  }
+
+-- | A 'ProgressReporting' that sends the WorkDone Begin and End notifications
+--   synchronously. Progress notifications are sent from a sampling thread.
```

Comment:
Perhaps a line of comment saying why we want this (namely testing), and the guarantee it providers (you will always get the messages)